Clarify camera errors and warnings with gphoto2 and DSLRs

When a camera fails there is a misleading and confusing set of error messages and warnings that appear in the main terminal screen. These errors/warnings should be more user friendly. (Note that this ticket is not for fixing the actual error, which is more involved).

```
E 08-08 11:04:36.711 InvalidSystemCommand: File: /home/panoptes/images/Toi1333B/f6eb3d/20250808T092713/20250808T110436.cr2
 err: Command '['/usr/bin/dcraw', '-t', '0', '-D', '-4', '/home/panoptes/images/Toi1333B/f6eb3d/20250808T092713/20250808T110436.cr2']' returned non-zero exit status 1.
E 08-08 11:04:36.901 _thread._ExceptHookArgs(exc_type=<class 'FileNotFoundError'>, exc_value=FileNotFoundError("Image file_path='/home/panoptes/images/Toi1333B/f6eb3d/20250808T092713/20250808T110436.fits' not found, cannot process."), exc_traceback=<trace
back object at 0xffff3a933200>, thread=<Thread(Thread-process_exposure-PAN007_f6eb3d_20250808T110436, started daemon 281473122824576)>)
E 08-08 11:05:21.523 Trying to process observation but missing file_path='/home/panoptes/images/Toi1333B/f6eb3d/20250808T092713/20250808T110436.cr2'
/home/panoptes/conda/envs/conda-pocs/lib/python3.12/site-packages/panoptes/utils/images/fits.py:466: UserWarning: No file exists at: /home/panoptes/images/Toi1333B/f6eb3d/20250808T092713/20250808T110436.cr2
  assert os.path.exists(fits_fname), warn(
Process Process-413:
Traceback (most recent call last):
  File "/home/panoptes/conda/envs/conda-pocs/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/home/panoptes/conda/envs/conda-pocs/lib/python3.12/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/panoptes/pocs/src/panoptes/pocs/observatory.py", line 520, in process_observation
    compressed_file_path = fits_utils.fpack(str(file_path))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/panoptes/conda/envs/conda-pocs/lib/python3.12/site-packages/panoptes/utils/images/fits.py", line 466, in fpack
    assert os.path.exists(fits_fname), warn(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: None
/home/panoptes/images/Toi1333B/f6eb3d/20250808T092713/20250808T110558.cr2: No such file or directory
```

--- a/src/panoptes/pocs/camera/camera.py
+++ b/src/panoptes/pocs/camera/camera.py
@@ -618,8 +618,8 @@ class AbstractCamera(PanBase, metaclass=ABCMeta):
                 self._is_observing_event.clear()
                 raise FileNotFoundError(f"Image {file_path=!r} not found, cannot process.")
         except (KeyError, FileNotFoundError) as e:
+            self.logger.error(f'Image {file_path!r} not found, cannot process.')
             self._is_observing_event.clear()
-            raise e
 
         # Do the camera specific processing.
         self.logger.debug(f'Starting FITS processing for {file_path}')

--- a/src/panoptes/pocs/camera/gphoto/base.py
+++ b/src/panoptes/pocs/camera/gphoto/base.py
@@ -270,11 +270,12 @@ class AbstractGPhotoCamera(AbstractCamera, ABC):  # pragma: no cover
 
     def _readout(self, filename, headers, *args, **kwargs):
         self.logger.debug(f'Reading Canon DSLR exposure for {filename=}')
+
         try:
             self.logger.debug(f"Converting CR2 -> FITS: {filename}")
             fits_path = cr2_utils.cr2_to_fits(filename, headers=headers, remove_cr2=False)
-        except TimeoutError:
-            self.logger.error(f'Error processing exposure for {filename} on {self}')
+        except Exception as err:
+            self.logger.error(f'Error processing exposure for {filename} on {self}: {err!r}')
         finally:
             self._readout_complete = True
 

--- a/src/panoptes/pocs/camera/gphoto/base.py
+++ b/src/panoptes/pocs/camera/gphoto/base.py
@@ -2,14 +2,14 @@ import re
 import subprocess
 import time
 from abc import ABC
+from pathlib import Path
 from typing import Dict, List, Union
 
+from panoptes.pocs.camera import AbstractCamera, get_gphoto2_cmd
 from panoptes.utils import error
 from panoptes.utils.images import cr2 as cr2_utils
 from panoptes.utils.serializers import from_yaml
 from panoptes.utils.utils import listify
-
-from panoptes.pocs.camera import AbstractCamera, get_gphoto2_cmd
 
 file_save_re = re.compile(r'Saving file as (.*)')
 
@@ -272,8 +272,11 @@ class AbstractGPhotoCamera(AbstractCamera, ABC):  # pragma: no cover
         self.logger.debug(f'Reading Canon DSLR exposure for {filename=}')
 
         try:
-            self.logger.debug(f"Converting CR2 -> FITS: {filename}")
-            fits_path = cr2_utils.cr2_to_fits(filename, headers=headers, remove_cr2=False)
+            if Path(filename).exists():
+                self.logger.debug(f"Converting CR2 -> FITS: {filename}")
+                fits_path = cr2_utils.cr2_to_fits(filename, headers=headers, remove_cr2=False)
+            else:
+                self.logger.warning(f"File {filename!r} not found, cannot process.")
         except Exception as err:
             self.logger.error(f'Error processing exposure for {filename} on {self}: {err!r}')
         finally:

--- a/src/panoptes/pocs/camera/gphoto/canon.py
+++ b/src/panoptes/pocs/camera/gphoto/canon.py
@@ -1,3 +1,5 @@
+from functools import lru_cache
+
 from astropy import units as u
 from panoptes.utils import error
 from panoptes.utils.time import current_time
@@ -141,6 +143,7 @@ class Camera(AbstractGPhotoCamera):
             return readout_args
 
     @classmethod
+    @lru_cache(maxsize=52)
     def get_shutterspeed_index(cls, seconds: float, return_minimum: bool = False):
         """Looks up the appropriate shutterspeed setting for the given seconds.
 

--- a/src/panoptes/pocs/camera/gphoto/canon.py
+++ b/src/panoptes/pocs/camera/gphoto/canon.py
@@ -1,6 +1,8 @@
 from functools import lru_cache
 
 from astropy import units as u
+from panoptes.utils.error import PanError
+
 from panoptes.utils import error
 from panoptes.utils.time import current_time
 from panoptes.utils.utils import get_quantity_value
@@ -138,6 +140,7 @@ class Camera(AbstractGPhotoCamera):
             self.command(cmd_args, check_exposing=False)
         except error.InvalidCommand as e:
             self.logger.warning(e)
+            raise PanError(f'Problem taking picture with {self.name}: {e}')
         else:
             readout_args = (filename, header)
             return readout_args

--- a/src/panoptes/pocs/observatory.py
+++ b/src/panoptes/pocs/observatory.py
@@ -527,7 +527,7 @@ class Observatory(PanBase):
                     metadata['filepath'] = compressed_file_path
                     self.logger.debug(f'Compressed {compressed_file_path}')
                 except (FileNotFoundError, AssertionError) as e:
-                    self.logger.warning(f'Problem compressing, file  not found {file_path=}: {e!r}')
+                    self.logger.warning(f'Problem compressing, file not found {file_path=}: {e!r}')
 
             bucket_name = self.get_config('panoptes_network.buckets.upload')
             # Get the images directory.

--- a/src/panoptes/pocs/observatory.py
+++ b/src/panoptes/pocs/observatory.py
@@ -6,10 +6,6 @@ from astropy import units as u
 from astropy.coordinates import get_body
 from astropy.io.fits import setval
 from contextlib import suppress
-from panoptes.utils import error, images as img_utils
-from panoptes.utils.images import fits as fits_utils
-from panoptes.utils.time import CountdownTimer, current_time, flatten_time
-from panoptes.utils.utils import get_quantity_value
 from pathlib import Path
 from typing import Dict, Optional
 
@@ -25,6 +21,10 @@ from panoptes.pocs.scheduler.observation.compound import Observation as Compound
 from panoptes.pocs.scheduler.scheduler import BaseScheduler
 from panoptes.pocs.utils.cloud import upload_image as image_uploader
 from panoptes.pocs.utils.location import create_location_from_config
+from panoptes.utils import error, images as img_utils
+from panoptes.utils.images import fits as fits_utils
+from panoptes.utils.time import CountdownTimer, current_time, flatten_time
+from panoptes.utils.utils import get_quantity_value
 
 
 class Observatory(PanBase):
@@ -278,8 +278,10 @@ class Observatory(PanBase):
         wait_timer = CountdownTimer(120, name='FinishCameraExposureWait')
         while self.current_observation and any(cam.is_observing for cam in self.cameras.values()):
             if wait_timer.expired():
-                self.logger.warning("Timeout waiting for cameras to finish observing, "
-                                    "proceeding with the parking of the mount.")
+                self.logger.warning(
+                    "Timeout waiting for cameras to finish observing, "
+                    "proceeding with the parking of the mount."
+                    )
                 break
 
             self.logger.debug(f"Waiting for cameras to finish observing, please be patient...{wait_timer}")
@@ -403,7 +405,11 @@ class Observatory(PanBase):
         for cam_name, camera in self.cameras.items():
             self.logger.debug(f"Exposing for camera: {cam_name}")
             # Don't block in this call but handle blocking below.
-            camera.take_observation(self.current_observation, headers=headers, blocking=False)
+            try:
+                camera.take_observation(self.current_observation, headers=headers, blocking=False)
+            except Exception as e:
+                self.logger.warning(f"Can't take observation for camera {cam_name}. Removing the camera")
+                del self.cameras[cam_name]
 
         if blocking:
             cam = self.primary_camera

--- a/src/panoptes/pocs/observatory.py
+++ b/src/panoptes/pocs/observatory.py
@@ -1,18 +1,17 @@
-import os
-import time
 from collections import OrderedDict
-from contextlib import suppress
-from pathlib import Path
-from typing import Dict, Optional
 
 import numpy as np
+import os
 from astropy import units as u
 from astropy.coordinates import get_body
 from astropy.io.fits import setval
+from contextlib import suppress
 from panoptes.utils import error, images as img_utils
 from panoptes.utils.images import fits as fits_utils
 from panoptes.utils.time import CountdownTimer, current_time, flatten_time
 from panoptes.utils.utils import get_quantity_value
+from pathlib import Path
+from typing import Dict, Optional
 
 import panoptes.pocs.camera.fli
 from panoptes.pocs.base import PanBase
@@ -20,10 +19,10 @@ from panoptes.pocs.camera import AbstractCamera
 from panoptes.pocs.dome import AbstractDome
 from panoptes.pocs.images import Image
 from panoptes.pocs.mount.mount import AbstractMount
-from panoptes.pocs.scheduler.scheduler import BaseScheduler
 from panoptes.pocs.scheduler.field import Field
 from panoptes.pocs.scheduler.observation.base import Observation
 from panoptes.pocs.scheduler.observation.compound import Observation as CompoundObservation
+from panoptes.pocs.scheduler.scheduler import BaseScheduler
 from panoptes.pocs.utils.cloud import upload_image as image_uploader
 from panoptes.pocs.utils.location import create_location_from_config
 
@@ -491,9 +490,14 @@ class Observatory(PanBase):
             # Check for a FITS file of whatever file_path we have.
             if Path(file_path).with_suffix('.fits').exists():
                 file_path = Path(file_path).with_suffix('.fits').as_posix()
+            else:
+                # Give a warning and skip processing.
+                self.logger.warning(f'No FITS file found for processing: {file_path=}')
+                return
 
-            if Path(file_path).exists() is False:
+            if not Path(file_path).exists():
                 self.logger.error(f'Trying to process observation but missing {file_path=}')
+                return
 
             if metadata.get('status') == 'complete':
                 self.logger.debug(f'{image_id} has already been processed, skipping')
@@ -517,10 +521,13 @@ class Observatory(PanBase):
 
             if compress_fits:
                 self.logger.debug(f'Compressing {file_path=!r}')
-                compressed_file_path = fits_utils.fpack(str(file_path))
-                exposure.path = Path(compressed_file_path)
-                metadata['filepath'] = compressed_file_path
-                self.logger.debug(f'Compressed {compressed_file_path}')
+                try:
+                    compressed_file_path = fits_utils.fpack(str(file_path))
+                    exposure.path = Path(compressed_file_path)
+                    metadata['filepath'] = compressed_file_path
+                    self.logger.debug(f'Compressed {compressed_file_path}')
+                except (FileNotFoundError, AssertionError) as e:
+                    self.logger.warning(f'Problem compressing, file  not found {file_path=}: {e!r}')
 
             bucket_name = self.get_config('panoptes_network.buckets.upload')
             # Get the images directory.

